pollen.8
.TH pollen 1 "07 February 2013" pollen "pollen server"
.SH NAME
\fBpollen\fP \- an Entropy-as-a-Service web server

.SH SYNOPSIS
\fBpollen\fP [OPTION]...

.SH OPTIONS

\fB-http-port\fP - the HTTP port on which to listen and serve cleartext responses; use "" to disable; default is "80"

\fB-https-port\fP - the HTTPS port on which to listen and serve encrypted, TLS responses; use "" to disable; default is "443"

\fB-device\fP - the device to use for reading and writing random data; default is \fI/dev/urandom\fP

\fB-bytes\fP - the size, in bytes, to transmit and receive each time to peers or neighbors listening in the pool; default is 64

\fB-cert\fP - the path to the TLS certificate; default is \fI/etc/pollen/cert.pem\fP

\fB-key\fP - the path to the TLS key; default is \fI/etc/pollen/key.pem\fP

.SH DESCRIPTION
\fBpollen\fP is an Entropy-as-a-Service web server, providing random seeds over a TLS encrypted connection.

All requests are serviced over HTTPS, using the key at \fI/etc/pollen/key.pem\fP and the cert at \fI/etc/pollen/cert.pem\fP.

Some configuration options are available to the system administrator in \fI/etc/default/pollen\fP.

.SH SEE ALSO
\fBpollen\fP(1), \fBpollinate\fP(1)

.SH AUTHOR
This manpage and the utility were written by Dustin Kirkland <kirkland@ubuntu.com> for Ubuntu systems (but may be used by others).  Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 3 published by the Free Software Foundation.

On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL, or on the web at \fIhttp://www.gnu.org/licenses/gpl.txt\fP.


debian/changelog
pollen (4.7) unreleased; urgency=low

  * UNRELEASED

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 18 Feb 2014 23:18:59 -0600

pollen (4.6-0ubuntu1) trusty; urgency=low

  [ Caleb Spare ]
  * pollen.go:
    - Require the challenge query-string param to be provided
    - don't create the random device, if it doesn't exist

  [ Dustin Kirkland ]
  * pollen_test.go:
    - update test to handle required challenge string
  * pollen.go:
    - incorporate feedback from Adam Langley
    - catch errors reading the random device
    - add a note as to why we're checksumming the random seed
    - update message when challenge empty

  [ Caleb Spare and Dustin Kirkland ]
  * debian/pollen.upstart, pollen.8, pollen.go:
    - Use flags rather than positional arguments, and plumb bytes argument
      through

  [ Dustin Kirkland and Matt Croydon ]
  * debian/pollen.default, debian/pollen.upstart, pollen.8, pollen.go:
    - add support for specifying the TLS cert and key as command line
      flags

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 18 Feb 2014 23:18:55 -0600

pollen (4.5-0ubuntu1) trusty; urgency=low

  [ Caleb Spare ]
  * pollen.go, pollen_test.go:
    - Bring naming in line with Go conventions
    - Use shorter parameter names for an http.HandlerFunc
    - Remove an unnecessary string conversion
    - Print useful error if wrong arguments are given rather than crashing
    - Don't ignore errors
    - Rename http[s]Port to http[s]Addr for accuracy
    - Handle errors starting the http servers
    - Change some naming in the test
    - Read from the provided device rather than always /dev/urandom

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 18 Feb 2014 14:50:52 -0600

pollen (4.4-0ubuntu1) trusty; urgency=low

  [ Casey Marshall ]
  * debian/control, debian/rules, Makefile, pollen.go, pollen_test.go:
    - add unit tests for pollen server

  [ Dustin Kirkland ]
  * debian/pollen.lintian-overrides:
    - override expected Lintian gripes

 -- Dustin Kirkland <kirkland@ubuntu.com>  Mon, 17 Feb 2014 12:51:51 -0600

pollen (4.3-0ubuntu1) trusty; urgency=low

  * check_pollen:
    - ensure that the nagios script uses the -r|--reseed option

 -- Dustin Kirkland <kirkland@ubuntu.com>  Mon, 17 Feb 2014 09:38:51 -0600

pollen (4.2-0ubuntu1) trusty; urgency=low

  * pollen.go:
    - remove redundant line
  * README:
    - remove deprecated bit of documentation

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 11 Feb 2014 18:04:08 -0600

pollen (4.1-0ubuntu1) trusty; urgency=low

  * debian/control:
    - build on amd64 and i386 only
    - these are the only builds I've been able to confirm when building
      with golang-go
    - note that this undoes the fix for LP: #1274074, but that's the
      way it has to be, until either golang-go supports more architectures
      or gccgo doesn't suck

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 11 Feb 2014 10:11:19 -0600

pollen (4.0-0ubuntu1) trusty; urgency=low

  * check_pollen, debian/control, debian/copyright,
    debian/pollinate.default, debian/pollinate.install,
    debian/pollinate.manpages, debian/pollinate.postinst,
    debian/pollinate.postrm, debian/pollinate.upstart,
    entropy.ubuntu.com.pem, INSTALL, Makefile, pollinate, pollinate.1:
    - split pollen and pollinate into separate projects and packages
    - re-enable the pollen build

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 11 Feb 2014 09:40:21 -0600

pollen (3.17-0ubuntu1) trusty; urgency=low

  * pollinate:
    - improve kernel debug info
  * debian/control, debian/pollen.install, Makefile:
    - TEMPORARILY disabling the building of pollen, until
      either gccgo or golang-go get promoted to main
    - this should be reverted as soon as a go compiler
      is available as a build dep

 -- Dustin Kirkland <kirkland@ubuntu.com>  Mon, 10 Feb 2014 14:16:08 -0600

pollen (3.16-0ubuntu1) trusty; urgency=low

  * pollinate:
    - minor standardization of the user agent string

 -- Dustin Kirkland <kirkland@ubuntu.com>  Wed, 05 Feb 2014 13:57:42 +0200

pollen (3.15-0ubuntu1) trusty; urgency=low

  * debian/control: LP: #1274074
    - build on any architecure, now that we build with gccgo

 -- Dustin Kirkland <kirkland@ubuntu.com>  Wed, 05 Feb 2014 12:31:20 +0200

pollen (3.14-0ubuntu1) trusty; urgency=low

  * debian/pollinate.postinst:
    - fix order of operations, packaging breakage

 -- Dustin Kirkland <kirkland@ubuntu.com>  Wed, 05 Feb 2014 11:34:36 +0200

pollen (3.13-0ubuntu1) trusty; urgency=low

  * README:
    - fix more minor typos
    - explain "did some work"
  * debian/rules, Makefile:
    - fix the build for gccgo
    - must use the -g parameter
    - don't strip binaries
    - these are ugly, but are the result of gccgo vs golang-go
  * pollinate:
    - remove unused variable $cmd
  * debian/pollinate.upstart:
    - our upstart job should start on starting cloud-init, to ensure that
      we get run before generating SSH keys
  * debian/pollinate.install, debian/pollinate.postrm, pollen.go,
    pollinate, pollinate.cron.d, README:
    - drop the tag and cronjob per feedback from sarnold in the code audit
      in LP: #1246098
  * debian/pollinate.default, pollinate:
    - add helpful debug info to user agent, similar to chrome and firefox,
  * debian/pollinate.postinst, debian/pollinate.postrm,
    debian/pollinate.upstart, pollinate, pollinate.1:
    - use a pollinate user, rather than the daemon user
    - by default, only run pollinate once per system instantiation
    - offer reseeding as an option, though
  * debian/control:
    - need to depend on adduser

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 04 Feb 2014 11:51:22 +0200

pollen (3.12-0ubuntu1) trusty; urgency=low

  * README:
    - minor documentation feedback from Kees Cook
    - note that pollen servers can of course be run internally
  * debian/control:
    - clean up package descriptions a bit

 -- Dustin Kirkland <kirkland@ubuntu.com>  Tue, 28 Jan 2014 22:16:10 +0000


pollen.go
/*

pollen: Entropy-as-a-Server web server

  Copyright (C) 2012-2013 Dustin Kirkland <dustin.kirkland@gmail.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as published by
  the Free Software Foundation, version 3 of the License.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

package main

import (
	"crypto/sha512"
	"flag"
	"fmt"
	"io"
	"log/syslog"
	"net/http"
	"os"
	"time"
)

var (
	httpPort  = flag.String("http-port", "80", "The HTTP port on which to listen")
	httpsPort = flag.String("https-port", "443", "The HTTPS port on which to listen")
	device    = flag.String("device", "/dev/urandom", "The device to use for reading and writing random data")
	size      = flag.Int("bytes", 64, "The size in bytes to transmit and receive each time")
	cert      = flag.String("cert", "/etc/pollen/cert.pem", "The full path to cert.pem")
	key       = flag.String("key", "/etc/pollen/key.pem", "The full path to key.pem")
	log *syslog.Writer
	dev *os.File
)

func handler(w http.ResponseWriter, r *http.Request) {
	challenge := r.FormValue("challenge")
	if challenge == "" {
		http.Error(w, "Please use the pollinate client.  'sudo apt-get install pollinate' or download from: https://bazaar.launchpad.net/~pollinate/pollinate/trunk/view/head:/pollinate", http.StatusBadRequest)
		return
	}
	checksum := sha512.New()
	io.WriteString(checksum, challenge)
	challengeResponse := checksum.Sum(nil)
	var err error
	_, err = dev.Write(challengeResponse)
	if err != nil {
		/* Non-fatal error, but let's log this to syslog */
		log.Err(fmt.Sprintf("Cannot write to random device at [%v]", time.Now().UnixNano()))
	}
	log.Info(fmt.Sprintf("Server received challenge from [%s, %s] at [%v]", r.RemoteAddr, r.UserAgent(), time.Now().UnixNano()))
	data := make([]byte, *size)
	_, err = io.ReadFull(dev, data)
	if err != nil {
		/* Fatal error for this connection, if we can't read from device */
		log.Err(fmt.Sprintf("Cannot read from random device at [%v]", time.Now().UnixNano()))
		http.Error(w, "Failed to read from random device", 500)
		return
	}
	checksum.Write(data[:*size])
	/* The checksum of the bytes from /dev/urandom is simply for print-ability, when debugging */
	seed := checksum.Sum(nil)
	fmt.Fprintf(w, "%x\n%x\n", challengeResponse, seed)
	log.Info(fmt.Sprintf("Server sent response to [%s, %s] at [%v]", r.RemoteAddr, r.UserAgent(), time.Now().UnixNano()))
}

func init() {
	var err error
	log, err = syslog.New(syslog.LOG_ERR, "pollen")
	if err != nil {
		fatalf("Cannot open syslog:", err)
	}
	dev, err = os.OpenFile(*device, os.O_RDWR, 0)
	if err != nil {
		fatalf("Cannot open device: %s\n", err)
	}
}

func main() {
	flag.Parse()
	defer dev.Close()
	if *httpPort == "" && *httpsPort == "" {
		fatal("Nothing to do if http and https are both disabled")
	}
	httpAddr := fmt.Sprintf(":%s", *httpPort)
	httpsAddr := fmt.Sprintf(":%s", *httpsPort)
	http.HandleFunc("/", handler)
	go func() {
		fatal(http.ListenAndServe(httpAddr, nil))
	}()
	fatal(http.ListenAndServeTLS(httpsAddr, *cert, *key, nil))
}

func fatal(args ...interface{}) {
	log.Crit(fmt.Sprint(args...))
	fmt.Fprint(os.Stderr, args...)
	os.Exit(1)
}

func fatalf(format string, args ...interface{}) {
	log.Emerg(fmt.Sprintf(format, args...))
	fmt.Fprintf(os.Stderr, format, args...)
	os.Exit(1)
}
